django/db/transaction.py

```diff
@@ -158,16 +158,27 @@ class Atomic(ContextDecorator):
 
     Since database connections are thread-local, this is thread-safe.
 
+    An atomic block can be tagged as durable. This tells django to raise a
+    RuntimeError if this block is nested within another atomic block. This
+    guarantees that database changes in a durable block are committed to
+    the database when the block exists without error.
+
     This is a private API.
     """
 
-    def __init__(self, using, savepoint):
+    def __init__(self, using, savepoint, durable):
         self.using = using
         self.savepoint = savepoint
+        self.durable = durable
 
     def __enter__(self):
         connection = get_connection(self.using)
 
+        if self.durable and connection.ensure_durability and connection.in_atomic_block:
+            raise RuntimeError(
```

Comment:
Would a [ProgrammingError](https://docs.djangoproject.com/en/3.1/ref/exceptions/#django.db.ProgrammingError) be better?

Reply:
[db-api pep](https://www.python.org/dev/peps/pep-0249/#programmingerror) says:

> Exception raised for programming errors, e.g. table not found or already exists, syntax error in the SQL statement, wrong number of parameters specified, etc. It must be a subclass of DatabaseError.

I think it's kind of meant for problems in programming the SQL, rather than python-level problems?

Reply:
I'm not sure - it's not clear to me what `ProgrammingError` should cover.